src/persist-client/src/impl/gc.rs

```diff
@@ -85,38 +90,93 @@ pub struct GarbageCollector {
+                }
+
+                let merged_requests = gc_completed_senders.len() - 1;
+                if merged_requests > 0 {
+                    debug!(
+                        "Merged {} gc requests together for shard {}",
+                        merged_requests, shard_id
+                    );
+                }
+
+                let consolidated_gc_req = GcReq {
+                    shard_id,
+                    old_seqno_since: oldest_seqno_since,
+                    new_seqno_since: newest_seqno_since,
+                };
+
+                let consensus = Arc::clone(&consensus);
+                let blob = Arc::clone(&blob);
+                let metrics = Arc::clone(&metrics);
+
+                let gc_span = debug_span!(parent: None, "gc_and_truncate", shard_id=%req.shard_id);
+                gc_span.follows_from(&Span::current());
 
-    pub fn gc_and_truncate_background(&self, req: GcReq) -> JoinHandle<()> {
-        // Spawn GC in a background task, so the state change that triggered it
-        // isn't blocked on it. Note that unlike compaction, GC maintenance
-        // intentionally has no "skipped" heuristic.
-        let gc_span = debug_span!(parent: None, "gc_and_truncate", shard_id=%req.shard_id);
-        gc_span.follows_from(&Span::current());
-
-        let consensus = Arc::clone(&self.consensus);
-        let blob = Arc::clone(&self.blob);
-        let metrics = Arc::clone(&self.metrics);
-        mz_ore::task::spawn(
-            || "persist::gc_and_truncate",
-            async move {
                 let start = Instant::now();
                 metrics.gc.started.inc();
-                Self::gc_and_truncate(consensus, blob, &metrics, req).await;
+                Self::gc_and_truncate(consensus, blob, &metrics, consolidated_gc_req)
+                    .instrument(gc_span)
+                    .await;
                 metrics.gc.finished.inc();
                 metrics.gc.seconds.inc_by(start.elapsed().as_secs_f64());
+
+                // inform all callers who enqueued GC reqs that their work is complete
+                for sender in gc_completed_senders {
+                    let _ = sender.send(());
+                }
             }
-            .instrument(gc_span),
-        )
+        });
+
+        GarbageCollector {
+            sender: gc_req_sender,
+        }
+    }
+
+    /// Enqueues a [GcReq] to be consumed by the GC background task when available.
+    ///
+    /// Returns a future that indicates when GC has cleaned up to at least [GcReq::new_seqno_since]
+    pub fn gc_and_truncate_background(
+        &self,
+        req: GcReq,
+    ) -> impl Future<Output = Result<(), RecvError>> {
+        let (gc_completed_sender, gc_completed_receiver) = oneshot::channel();
+        let new_gc_sender = self.sender.clone();
+        let shard_id = req.shard_id.clone();
+        let new_seqno_since = req.new_seqno_since.clone();
+        Handle::current().spawn_named(
```

Comment:
oh in case the channel is full? maybe leave that as a comment

Reply:
it's much grosser than that -- the channel is async, so you must `await` in order to send. But our function `gc_and_truncate_background` is sync, so it can't await, and because the function is sync but usually called by an async function, we can't use `Handle::current().block_on(...)` because then we'll block the thread that's driving the future.

One of my attempts was making `gc_and_truncate_background` async, but that also had some knock on effects that felt very awkward to work with that I don't remember the specifics of now.

Reply:
I might revisit this though, maybe it's easier just to make it all async in the first place now that the rest of the pieces are settled